[camera] Add ability to concurrently record and stream video

adam-harwood opened the pull request:
This PR adds the ability to optionally also enable streaming when starting recording of a video. This feature is useful, for example, in applications that use image recognition. The customer will want some immediate feedback as the video is being recorded (where an ML model might be run on the captured images), as well as to upload the video to a server at the end for final server-side processing.

Functionally, the new optional named parameter on `startVideoRecording` will combine the existing implementation with that of `startImageStream`. Besides this, all functionality remains the same. For existing code, that does not use the new parameter, there is no change to the code path.

I have added unit tests to the higher-level dart abstractions, added integration tests for the underlying android and iOS implementations, and tested manually on Android and iOS devices. I have updated `camera_web` to throw an exception if the user attempts to enable streaming when recording (given that streaming is not yet supported).

Closes https://github.com/flutter/flutter/issues/83634

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [X] I signed the [CLA].
- [X] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [X] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making, or this PR is [test-exempt].
- [X] All existing and new tests are passing.

adam-harwood commented:
Looks like I need to update `camera_windows` as well. Will do.

adam-harwood commented:
One of the checks is failing because it doesn't want the PR to include both changes to `camera_platform_interface` and the implementations. Can you confirm if that's the case (and that you're happy with the changes in `camera_platform_interface`)? If so then I'll split this out into 2 PRs.

loic-sharma commented on packages/camera/camera_windows/lib/camera_windows.dart:
```diff
@@ -217,7 +217,13 @@ class CameraWindows extends CameraPlatform {
+    CameraImageStreamOptions? streamOptions
   }) async {
+    if (streamCallback != null || streamOptions != null) {
+      throw UnimplementedError('Streaming is not currently supported on web');
```
```suggestion
      throw UnimplementedError('Streaming is not currently supported on Windows');
```

hellohuanlin commented on packages/camera/camera_avfoundation/ios/Classes/FLTCam.h:
```diff
@@ -50,6 +50,8 @@ NS_ASSUME_NONNULL_BEGIN
 - (void)close;
 - (void)startVideoRecordingWithResult:(FLTThreadSafeFlutterResult *)result;
+- (void)startVideoRecordingWithResult:(FLTThreadSafeFlutterResult *)result
+                       streamCallback:(nullable NSObject<FlutterBinaryMessenger> *)messenger;
```
maybe `messengerForStreaming`? this should not be named `streamCallback` since it's not a callback block.

adam-harwood replied:
Will do.

hellohuanlin commented on packages/camera/camera_platform_interface/lib/src/method_channel/method_channel_camera.dart:
```diff
@@ -283,13 +292,17 @@ class MethodChannelCamera extends CameraPlatform {
+    return _frameStreamController!.stream;
+  }
+
+  void _installStreamController({Function()? onListen}) {
```
In `startVideoRecording` Can we directly pass `streamCallback` into `_installStreamController()` and make it nonNull? A no-op onListen seems a bad state to me.

adam-harwood replied:
Not sure I totally understand. `streamCallback` is the callback which receives each frame that is processed. `onListen` is different, it's a callback invoked once when the stream is listened on. In the case of `onStreamedFrameAvailable` it needs to call the native `startImageStream` to begin the stream, but on the `startVideoRecording` the camera stream has already been initialised during the call to the native `startVideoRecording`. Therefore, in the case of `startVideoRecording`, `onListen` should be a no-op.

I could make onListen non-null, and have it up to the caller to default it to a no-op function, but that seems to put more burden on the caller. Is that preferable?

hellohuanlin replied:
I meant do we want to have `onListen` for both use cases? It's a bit odd to have one streaming API with onListen, but the other streaming API without.

adam-harwood replied:
`onListen` isn't part of the public API though. It's an implementation detail inside this class. Nobody using the camera plugin will have to deal with it at all. The only reason that `onListen` is nullable is that, in the case where they want to both stream and record, the initialization is another section of the (internal) code. `StreamController` itself is a class external to the camera plugin and can't be easily modified.

Given that the method is private and internal to the class, is the difference still a concern?

hellohuanlin commented on packages/camera/camera_avfoundation/lib/src/avfoundation_camera.dart:
```diff
@@ -248,14 +248,23 @@ class AVFoundationCamera extends CameraPlatform {
+
+    if (streamCallback != null) {
+      _installStreamController();
+      _frameStreamController!.stream.listen(streamCallback);
```
can you comment out why _frameStreamController must not be nil?

adam-harwood replied:
Good point. I will actually change `_installStreamController` to return the controller, and then call listen on that. Will make the code easier to read, and remove the need for a `!`.

hellohuanlin commented on packages/camera/camera_avfoundation/lib/src/avfoundation_camera.dart:
```diff
@@ -291,13 +300,17 @@ class AVFoundationCamera extends CameraPlatform {
+  void _installStreamController({Function()? onListen}) {
     _frameStreamController = StreamController<CameraImageData>(
-      onListen: _onFrameStreamListen,
+      onListen: onListen ?? () {},
```
when do we want a no-op `onListen`

adam-harwood replied:
Hopefully addressed in [this comment](https://github.com/flutter/plugins/pull/6290/#discussion_r957914180), but let me know if not.

hellohuanlin reviewed (commented):
I only reviewed the iOS part

adam-harwood commented:
Thanks for the feedback @hellohuanlin , I'll wait for some feedback on the other modules as well and then address everything you raised.

camsim99 commented on packages/camera/camera_android/lib/src/android_camera.dart:
```diff
@@ -291,13 +300,17 @@ class AndroidCamera extends CameraPlatform {
+  void _installStreamController({Function()? onListen}) {
     _frameStreamController = StreamController<CameraImageData>(
-      onListen: _onFrameStreamListen,
+      onListen: onListen ?? () {},
```
Consider https://github.com/flutter/plugins/pull/6290/files#r952963757 here as well.

adam-harwood replied:
Replied in [this comment](https://github.com/flutter/plugins/pull/6290/#discussion_r957914180), I don't think it applies unless I'm misunderstanding something.

camsim99 reviewed (commented):
Reviewed Android part!

adam-harwood commented on packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/Camera.java:
```diff
@@ -524,6 +524,21 @@ private void refreshPreviewCaptureSession(
+    }
+
+    createCaptureSession(
+        CameraDevice.TEMPLATE_RECORD, successCallback, surfaces.toArray(new Surface[0]));
```
This isn't actually calling the deprecated method, it's referencing a different internal [createCaptureSession](https://github.com/flutter/plugins/blob/8aa5225c49f6ef4cbdda096f62bc704a1c829228/packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/Camera.java#L392) method that isn't deprecated. That method then calls the non-deprecated [CameraDevice.createCaptureSession](https://github.com/flutter/plugins/blob/8aa5225c49f6ef4cbdda096f62bc704a1c829228/packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/Camera.java#L481).

camsim99 replied:
Gotcha! Thanks for the clarification.

adam-harwood commented on packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/Camera.java:
```diff
@@ -744,28 +759,20 @@ private void unlockAutoFocus() {
   }
 
   public void startVideoRecording(@NonNull Result result) {
```
Looks like nothing is calling it, will remove.

adam-harwood commented on packages/camera/camera/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+## 0.11.0
+
+* Implement option to also stream when recording a video.
```
Will do.

adam-harwood commented on packages/camera/camera_android/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+## 0.11.0
+
+* Implement option to also stream when recording a video.
```
Will do.

loic-sharma commented on packages/camera/camera_windows/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+## NEXT
+
+* Update to latest camera platform interface, and fail if user attempts to use streaming with recording (since streaming is currently unsupported on web).
```
Please use the present tense for release notes. Also `web` should be Windows here.

```suggestion
* Updates to latest camera platform interface but fails if user attempts to use streaming with recording (since streaming is currently unsupported on Windows).
```

hellohuanlin commented:
Saw a few updates in my email. Just double check - let me know when it's ready for a second review.

adam-harwood commented:
> Saw a few updates in my email. Just double check - let me know when it's ready for a second review.

Yep, it's ready @hellohuanlin , thanks :)

hellohuanlin commented on packages/camera/camera_avfoundation/ios/Classes/FLTCam.h:
```diff
@@ -50,6 +50,8 @@ NS_ASSUME_NONNULL_BEGIN
 - (void)close;
 - (void)startVideoRecordingWithResult:(FLTThreadSafeFlutterResult *)result;
+- (void)startVideoRecordingWithResult:(FLTThreadSafeFlutterResult *)result
+                messengerForStreaming:(nullable NSObject<FlutterBinaryMessenger> *)messenger;
```
can you add a comment saying that the optional param is for concurrent streaming during video recording

adam-harwood replied:
Sure, done.

hellohuanlin commented on packages/camera/camera_avfoundation/lib/src/avfoundation_camera.dart:
```diff
@@ -248,14 +248,22 @@ class AVFoundationCamera extends CameraPlatform {
     );
+
+    if (streamCallback != null) {
+      _installStreamController().stream.listen(streamCallback);
```
This is doing 2 complex things in 1 line, making it a bit hard to understand. Can you split it?

adam-harwood replied:
No problem, done.

hellohuanlin commented on packages/camera/camera_avfoundation/lib/src/avfoundation_camera.dart:
```diff
@@ -291,13 +299,19 @@ class AVFoundationCamera extends CameraPlatform {
       onCancel: _onFrameStreamCancel,
     );
-    return _frameStreamController!.stream;
+    return _frameStreamController!;
```
Why are you returning an ivar here? (ivars should be accessible from the caller already).

adam-harwood replied:
By returning the variable we remove the need for the caller to _know_ that it is operating on an ivar. This makes refactoring simpler: it could later be changed from being an instance variable to a local variable (or read from another instance), or a subclass could override the implementation, or the return value mocked, and everything will work as before.

hellohuanlin replied:
this would make this function unnecessarily complex, because the side effect of assigning the ivar is not obvious to the caller. I recommend just creating the controller and return it. Then update the ivar in caller site. Something like: 

```
StreamController<CameraImageData> _createStreamController(
      {Function()? onListen})

// caller
_frameStreamController = _createStreamController(...)
```

adam-harwood replied:
As you wish, I have made that change. Ready to review.

adam-harwood commented:
@hellohuanlin this should be ready for review again.

jmagman commented:
@hellohuanlin can you take another look at this, if your changes have all been addressed?

hellohuanlin commented on packages/camera/camera_avfoundation/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+## 0.10.0
+
+* Implement option to also stream when recording a video.
```
reminder if this is not done yet

adam-harwood replied:
Sorry, that missed my todo list. Fixed now.

hellohuanlin reviewed (approved):
ios looks good

jmagman commented:
@loic-sharma this still needs Windows review.

@ditman you were marked as the camera_web codeowner, did you remove yourself as a reviewer?  Not sure how that disappeared.

loic-sharma reviewed (commented):
Windows changes LGTM

loic-sharma reviewed (approved):
Windows changes LGTM

cbracken reviewed (approved):
lgtm for avfoundation

ditman reviewed (approved):
@jmagman I was removed [here](https://github.com/flutter/plugins/pull/6290#event-7283525384).

I'm fine with the changes for the web version (it doesn't support streaming, and they're consistent (throw) with other platforms).

bparrishMines reviewed (approved):
LGTM app-facing changes

stuartmorgan commented on packages/camera/camera_platform_interface/lib/src/platform_interface/camera_platform.dart:
```diff
@@ -125,12 +125,17 @@ abstract class CameraPlatform extends PlatformInterface {
   /// With [maxVideoDuration] set the video is returned in a [VideoRecordedEvent]
   /// through the [onVideoRecordedEvent] stream when the set duration is reached.
-  Future<void> startVideoRecording(int cameraId, {Duration? maxVideoDuration}) {
+  Future<void> startVideoRecording(int cameraId,
```
This is a breaking change, so can't be made this way. Please see https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changing-platform-interface-method-parameters

(Sorry @bparrishMines I meant to tag you in for app-facing + platform interface, not just for app-facing.)

adam-harwood replied:
Cool, I [figured this was the case](https://github.com/flutter/plugins/pull/6290#issuecomment-1220092690) but was just waiting for some confirmation. I'll split this out into a separate PR and then update this one accordingly.

adam-harwood replied:
PR created: https://github.com/flutter/plugins/pull/6550

adam-harwood replied:
This has been merged now and this PR updated accordingly.

adam-harwood commented on packages/camera/camera/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+## 0.11.0
```
I wasn't aware of the rules for Dart semver, so when it's < 1.0 then a minor version change is considered breaking? Given that, I think I should make this 0.10.1? Or should it be 0.10.0+2?

stuartmorgan replied:
This is new API that's non-breaking, which is a minor bump in semver. Since Dart shifts everything over one for pre-1.0 it would be 0.10.1.

jmagman commented:
> last line of pub output: "Because camera_web depends on camera_platform_interface ^2.3.1 which doesn't match any versions, version solving failed."


@stuartmorgan this might need some guidance landing the last few changes, but it looks pretty much ready?

adam-harwood commented:
> > last line of pub output: "Because camera_web depends on camera_platform_interface ^2.3.1 which doesn't match any versions, version solving failed."
> 
> @stuartmorgan this might need some guidance landing the last few changes, but it looks pretty much ready?

Thanks @jmagman , currently I'm waiting on https://github.com/flutter/plugins/pull/6666 to get merged so that I can fix those errors you pointed out. Just some sequencing to be done.

adam-harwood commented:
I've updated this now with the changes from #6550 and #6666 . I have removed the changes to the `camera` package, since they are dependent on the changes in the various implementation packages (e.g. `camera_android`). Therefore, once this PR is merged I'll open another one to include the changes to `camera`.

faisalmushtaq007 commented:
has this part been merged?

adam-harwood commented:
> has this part been merged?

This PR? No. It's awaiting final review but should be ready to merge.

faisalmushtaq007 commented:
when will u give a version update?

faisalmushtaq007 commented:
@adam-harwood is their any set date for this merge and package update? Actually we have an app launch in coming days

adam-harwood commented:
> @adam-harwood is their any set date for this merge and package update? Actually we have an app launch in coming days

I don't know mate, I'm not on the Flutter team (I'm just an open source contributor) so I have no idea on what their merge and release schedule is. To be honest, this PR has been open since August 18 so I don't like your changes of having it merged and released in the next few days. I'd suggest coming up with a Plan B if I was you.

faisalmushtaq007 commented:
@stuartmorgan  @jmagman   why is the merging blocked??

faisalmushtaq007 commented:
@tarrinneal @gaaclarke @cyanglaz @GaryQian @jmagman @stuartmorgan Review was requested, please approve the changes!

adam-harwood commented:
> @adam-harwood cant we do anything about this blockage??

I can't sorry, like I said I don't work on the Flutter team and have no power here. I've been trying to get it merged for 4 months. I suggest you like for a plan B, possibly forking and distributing the code yourself.

stuartmorgan commented on packages/camera/camera_web/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+## 0.3.1
+
+* Update to latest camera platform interface, and fail if user attempts to use streaming with recording (since streaming is currently unsupported on web).
```
Nit: "Updates" and "fails" (per the CHANGELOG style guide linked from the PR checklist).

adam-harwood replied:
Done

stuartmorgan commented on packages/camera/camera_web/lib/src/camera_web.dart:
```diff
@@ -451,23 +451,33 @@ class CameraPlugin extends CameraPlatform {
+  @override
+  Future<void> startVideoCapturing(VideoCaptureOptions options) {
+    if (options.streamCallback != null || options.streamOptions != null) {
+      throw UnimplementedError('Streaming is not currently supported on web');
```
The unit tests should test that this throws as expected.

adam-harwood replied:
Done

stuartmorgan commented on packages/camera/camera_windows/lib/camera_windows.dart:
```diff
@@ -214,15 +214,24 @@ class CameraWindows extends CameraPlatform {
+  Future<void> startVideoCapturing(VideoCaptureOptions options) async {
+    if (options.streamCallback != null || options.streamOptions != null) {
+      throw UnimplementedError(
+          'Streaming is not currently supported on Windows');
```
The unit tests should test that this throws as expected.

stuartmorgan commented on packages/camera/camera_avfoundation/pubspec.yaml:
```diff
@@ -2,7 +2,7 @@ name: camera_avfoundation
 repository: https://github.com/flutter/plugins/tree/main/packages/camera/camera_avfoundation
 issue_tracker: https://github.com/flutter/flutter/issues?q=is%3Aissue+is%3Aopen+label%3A%22p%3A+camera%22
-version: 0.9.8+6
+version: 0.10.0
```
0.9.9, since this is not a breaking change.

adam-harwood replied:
Done

stuartmorgan reviewed (changes requested):
Just a few minor issues, then it's good to land! Sorry for the delay on my final review here.

stuartmorgan commented:
@faisalmushtaq007 Please don't spam people with pings.

adam-harwood commented:
@stuartmorgan this is ready for re-review.

stuartmorgan reviewed (approved):
LGTM!

stuartmorgan commented:
We'll need to revert this; the new Android integration test is failing. See https://github.com/flutter/plugins/pull/6796 for details. I should have pushed a trivial change to this PR to trigger the Android tests, but forgot.

@adam-harwood Are you able to reproduce that failure locally to debug it for an updated version of the PR?

adam-harwood commented:
> @adam-harwood Are you able to reproduce that failure locally to debug it for an updated version of the PR?

I ran them just now on an Android phone and cannot replicate the error you posted. I did get a different error though:
```
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following assertion was thrown running a test:
Guarded function conflict.
You must use "await" with all Future-returning test APIs.
The guarded method "pump" from class LiveTestWidgetsFlutterBinding was called from
package:flutter_test/src/binding.dart on line 841.
Then, the "expectLater" function was called from
file:///home/adam/dev/flutter_plugins/packages/camera/camera_android/example/integration_test/camera_test.dart
on line 275.
The first method (LiveTestWidgetsFlutterBinding.pump) had not yet finished executing at the time
that the second function (expectLater) was called. Since both are guarded, and the second was not a
nested call inside the first, the first must complete its execution before the second can be called.
Typically, this is achieved by putting an "await" statement in front of the call to the first.

When the first method (LiveTestWidgetsFlutterBinding.pump) was called, this was the stack:
#0      TestAsyncUtils.guard (package:flutter_test/src/test_async_utils.dart:69:54)
#1      LiveTestWidgetsFlutterBinding.pump (package:flutter_test/src/binding.dart:1653:27)
#2      TestWidgetsFlutterBinding._runTestBody (package:flutter_test/src/binding.dart:841:13)
<asynchronous suspension>

When the exception was thrown, this was the stack:
#0      TestAsyncUtils.guardSync (package:flutter_test/src/test_async_utils.dart:267:5)
#1      expectLater (package:flutter_test/src/widget_tester.dart:491:18)
#2      main.<anonymous closure>.<anonymous closure> (file:///home/adam/dev/flutter_plugins/packages/camera/camera_android/example/integration_test/camera_test.dart:275:9)
(elided 24 frames from dart:async and package:stack_trace)

The test description was:
  recording with image stream
════════════════════════════════════════════════════════════════════════════════════════════════════
01:46 +2 ~2 -1: recording with image stream [E]                                                                                                                            
  Test failed. See exception logs above.
  The test description was: recording with image stream
  
  'package:flutter_test/src/binding.dart': Failed assertion: line 1736 pos 12: '!_expectingFrame': is not true.
  dart:core                                                                  _AssertionError._throwNew
  package:flutter_test/src/binding.dart 1736:12                              LiveTestWidgetsFlutterBinding.postTest
  ===== asynchronous gap ===========================
  package:stream_channel                                                     _GuaranteeSink.add
  tmp/flutter_tools.TJHMNR/flutter_test_listener.FNKBUV/listener.dart 53:22  main.<fn>
  tmp/flutter_tools.TJHMNR/flutter_test_listener.FNKBUV/listener.dart 52:20  main.<fn>
```
which I can fix, but I think it's unlikely it will solve the problem you posted. Applying the fix locally, all tests are then green. I'm testing this on an Android device, what device do the automatic runs execute on @stuartmorgan ?

For reference, I'm running the tests with:
```
flutter_plugins/packages/camera/camera_android/example$ flutter test integration_test/camera_test.dart
```
Which I assume is correct.

stuartmorgan commented:
> which I can fix, but I think it's unlikely it will solve the problem you posted. Applying the fix locally, all tests are then green.

Could you post a new PR that starts as a revert of the revert, and then has a commit that fixes the test issue you found locally (so we can easily review just the diff)? I can then push a commit to that PR to trigger the tests and we can see how it looks in CI.

> I'm testing this on an Android device, what device do the automatic runs execute on @stuartmorgan ?

It looks like it's a Samsung Galaxy S9 and a Pixel 5.

> For reference, I'm running the tests with:

That should run the same tests. Unfortunately there's no easy way to run them exactly the same way as the CI runs them (we hope to add simulator tests at some point to avoid this disconnect), but that's usually not an issue.

adam-harwood commented:
> Could you post a new PR that starts as a revert of the revert, and then has a commit that fixes the test issue you found locally (so we can easily review just the diff)?

Sure @stuartmorgan : https://github.com/flutter/plugins/pull/6808
